### PR TITLE
Use `Arc<[Buffer]>` instead of raw `Vec<Buffer>` in `GenericByteViewA…

### LINKED ISSUE #6408: Make `StringViewArray::slice()` and `BinaryViewArray::slice()` faster
**Is your feature request related to a problem or challenge? Please describe what you are trying to do.**
While working on an upstream project https://github.com/apache/datafusion/pull/12092 which switched DataFusion to use `StringViewArray` rather than `StringArray`

When I did, one of the queries got much slower. 

Profiling revealed that the time difference was almost entirely explained by the time spent in `StringViewArray::slice()` 

Screenshot 2024-09-16 at 4 44 50 PM

Here is the flamegraph with `StringArray`:
flamegraph-main

Here is the same query with `StringViewArray`:
flamegraph-string-view

I am pretty sure the additional time is due to the time spent allocating / copying / deallocating the `Vec`s of buffers here:

https://github.com/apache/arrow-rs/blob/341ec357e74e897f50250930b44f453bce54a19a/arrow-array/src/array/byte_view_array.rs#L118

Where calling `slice` on a StringArray can be done with a few Arc increments. 

https://github.com/apache/arrow-rs/blob/3490639252294215c7ee05990d82b43e2cd097a6/arrow-array/src/array/byte_array.rs#L88-L91

**Describe the solution you'd like**
I would like `StringViewArray::slice` to be faster (aka don't allocate)

**Describe alternatives you've considered**

We can (and probably should) change DataFusion not to use slice in this case (I am working to file a ticket on this) but I think making `slice` faster / non allocating for `StringViewArray` will be useful in general

**Additional context**

### PR DESCRIPTION
**Pull request by ShiKaiWi**
…rray` for faster `slice`

# Which issue does this PR close?
Close #6408 

Closes #.

# Rationale for this change
In the `GenericByteViewArray`, the `buffers` field is a raw vector, leading to heap allocation when some methods, e.g. `clone`, `slice`. Using `Arc<[Buffer]>` instead of the raw `Vec<Buffer>` can avoid such heap allocation.

(TODO: add the benchmark results when the benchmark case is added)

# What changes are included in this PR?
Use `Arc<[Buffer]>` instead of the raw `Vec<Buffer>` as the type of `buffers` field of `GenericByteViewArray`.

# Are there any user-facing changes?
The signature of the method `GenericByteViewArray::new_unchecked` is changed from:
```
pub unsafe fn new_unchecked(
        views: ScalarBuffer<u128>,
        buffers: Vec<[Buffer]>,
        nulls: Option<NullBuffer>,
    ) -> Self;
```
to
```
pub unsafe fn new_unchecked(
        views: ScalarBuffer<u128>,
        buffers: impl Into<Arc<[Buffer]>>,
        nulls: Option<NullBuffer>,
    ) -> Self;
```

### CODE REVIEW

**Comment by tustvold** [comment]
Unfortunately the use of impl is still a breaking change as it could impact type inference, e.g if collecting an interator into the argument

File: arrow-array/src/array/byte_view_array.rs [hunk 1]
```diff
@@ -234,7 +234,7 @@ impl<T: ByteViewType + ?Sized> GenericByteViewArray<T> {
     }
 
     /// Deconstruct this array into its constituent parts
-    pub fn into_parts(self) -> (ScalarBuffer<u128>, Vec<Buffer>, Option<NullBuffer>) {
+    pub fn into_parts(self) -> (ScalarBuffer<u128>, Arc<[Buffer]>, Option<NullBuffer>) {
```
Comment (tustvold) [inline arrow-array/src/array/byte_view_array.rs:237]: This is also a breaking change

**Comment by alamb** [comment]
FWIW the reason "breaking change" is important is that it restricts when we can merge this PR: 

https://github.com/apache/arrow-rs/blob/master/CONTRIBUTING.md#breaking-changes

File: arrow-array/src/array/byte_view_array.rs [hunk 2]
```diff
@@ -114,7 +114,7 @@ use super::ByteArrayType;
 pub struct GenericByteViewArray<T: ByteViewType + ?Sized> {
     data_type: DataType,
     views: ScalarBuffer<u128>,
-    buffers: Vec<Buffer>,
+    buffers: Arc<[Buffer]>,
```
Comment (findepi) [inline arrow-array/src/array/byte_view_array.rs:117]: What's the rationale for Arc<[Buffer]> vs Vec<Arc<Buffer>>?

Comment (Dandandan) [reply arrow-array/src/array/byte_view_array.rs:117]: Cloning an `Arc` is relatively cheap (no allocation), cloning a `Vec` isn't.

Comment (findepi) [reply arrow-array/src/array/byte_view_array.rs:117]: i get it. However, if i understand correctly, `Arc<[Buffer]>` means the buffers can be passed around and shared only when they are within single slice, which can be limiting. For example, Can i merge two arrays, combining their `Arc<Buffer>` s without moving or cloning the buffers?

Comment (alamb) [reply arrow-array/src/array/byte_view_array.rs:117]: >  Can i merge two arrays, combining their Arc<Buffer> s without moving or cloning the buffers?

Comment (findepi) [reply arrow-array/src/array/byte_view_array.rs:117]: > you would have to create a new `Vec<Buffer`>

but that would prevent buffer sharing between two arrays, right?



> slicing / cloning is faster

cloning yes

slicing -- i didn't see it

Comment (alamb) [reply arrow-array/src/array/byte_view_array.rs:117]: I think the observation is that during `StringViewArray::slice`, the `slice` actually happens on the `view`s  -- the buffers (that the views can point at) must be copied

Here is the clone of `buffers`: https://docs.rs/arrow-array/53.1.0/src/arrow_array/array/byte_view_array.rs.html#385

**Comment by alamb** [comment]
> I wonder if kernels are blindly concatenating identical buffers together, instead of using something like Buffer::ptr_eq to avoid a new entry for the exact same buffer allocation?

What was happening in DataFusion was we had a Filter --> Coalesce chain and thus basically calling `concat` a few thousand  different input `RecordBatch` each with a few rows.

However, to your point, it may well be the case that the input RecordBatches shared the same underlying buffer so maybe the same buffer was being appended multiple times

@XiangpengHao do you remember if you looked for this ? (related to "Section 3.5: Buffer size tuning
" in https://www.influxdata.com/blog/faster-queries-with-stringview-part-two-influxdb/)

**Comment by XiangpengHao** [comment]
> I wonder if kernels are blindly concatenating identical buffers together, instead of using something like Buffer::ptr_eq to avoid a new entry for the exact same buffer allocation?

I think so: https://github.com/apache/arrow-rs/blob/def94a839236f3b04727a07c378668c9ada807f0/arrow-data/src/transform/mod.rs#L630-L637

>  Someone popped up on discord the other day reporting a StringViewArray with ~10k buffers, this would suggest something is likely off somewhere.

I have experienced this when loading string view from Parquet. If the parquet data has 10k buffers of string data, the string view will just hold them. Typically we should run a filter and then [gc](https://docs.rs/arrow/latest/arrow/array/type.StringViewArray.html#method.gc) it. This is handled in DF but if used out side DF users might need to do something similar like this: https://github.com/apache/datafusion/blob/c0ca4b4e449e07c3bcd6f3593fa31dd31ed5e0c5/datafusion/physical-plan/src/coalesce/mod.rs#L201-L221

In other words, if StringViewArray is constructed by us, it's very unlikely to have 10k buffers as we will exponentially grow the buffers ultil 2MB; so 10k buffers means ~20GB data

**Comment by alamb** [comment]
> In other words, if StringViewArray is constructed by us, it's very unlikely to have 10k buffers as we will exponentially grow the buffers ultil 2MB; so 10k buffers means ~20GB data

Good call -- @onursatici  has tracked down a related (perhaps the same) issue:

- https://github.com/apache/arrow-rs/issues/6780

**Comment by alamb** [comment]
All the discussion above not withstanding, I think the fact that this PR improves `slice` speed (by avoiding allocations) for GenericByteViewArray means it is still worth merging

File: arrow-array/src/array/byte_view_array.rs [hunk 3]
```diff
@@ -178,7 +178,7 @@ impl<T: ByteViewType + ?Sized> GenericByteViewArray<T> {
         Ok(Self {
             data_type: T::DATA_TYPE,
             views,
-            buffers,
+            buffers: buffers.into(),
```
Comment (tustvold) [inline arrow-array/src/array/byte_view_array.rs:181]: I think we should take `impl Into<Arc<[Buffer]>>` in this method as well

**Comment by tustvold** [comment]
> means it is still worth merging

Agreed

----

What do people think about extracting a newtype `ViewBuffers` or similar, much like we have `Fields`. This would allow us to change the internal representation down the line, whilst also potentially supporting member functions, etc...

**Comment by alamb** [comment]
> > means it is still worth merging
> 
> Agreed
> 
> What do people think about extracting a newtype `ViewBuffers` or similar, much like we have `Fields`. This would allow us to change the internal representation down the line, whilst also potentially supporting member functions, etc...

I think it would be a great idea. @ShiKaiWi  is that something you would be willing to try?

**Comment by ShiKaiWi** [comment]
> > > means it is still worth merging
> > 
> > 
> > Agreed
> > What do people think about extracting a newtype `ViewBuffers` or similar, much like we have `Fields`. This would allow us to change the internal representation down the line, whilst also potentially supporting member functions, etc...
> 
> I think it would be a great idea. @ShiKaiWi is that something you would be willing to try?

Sorry for the late reply (been a bit busy with work in the last few months). I'd love to try it on this weekend.

**Comment by alamb** [comment]
> Sorry for the late reply (I was a bit busy with work in the last few months). I'd love to try it on this weekend.

Thank @ShiKaiWi  -- I don't think there is any huge rush to get this feature in, so it would be great if you could do so ❤️

**Comment by alamb** [comment]
Converting to draft as I think there is some new planned work here.

**Comment by ctsk** [comment]
May I pick this up? It seems like it would be useful for DF :)

**Comment by alamb** [comment]
> May I pick this up? It seems like it would be useful for DF :)

Please do -- I agree I have seen this come up in some profiles

The other thing Arc<[Buffer]> might help with is reduce the cost of `clone()` ing StringViewArrays too

**Comment by Jefffrey** [comment]
Closing this PR as stale. Looks like it might be superseded by #9016 anyway.
